GetServices/handler.ts

```diff
@@ -40,11 +44,70 @@ type IGetServicesHandler = (
   auth: IAzureApiAuthorization
 ) => Promise<IGetServicesHandlerResult>;
 
+interface IFoldableResultIterator<T> {
+  readonly executeNext: (init: T) => Promise<Either<CosmosErrors, T>>;
+}
+
+function reduceResultIterator<A, B>(
+  i: AsyncIterator<ReadonlyArray<Either<Errors, A>>>,
+  f: (prev: B, curr: A) => B
+): IFoldableResultIterator<B> {
+  return {
+    executeNext: (init: B) =>
+      new Promise((resolve, reject) =>
+        i.next().then(
+          errorOrMaybeDocuments =>
+            errorOrMaybeDocuments.value().map(arrayOfMaybeDocs => {
+              if (rights(arrayOfMaybeDocs).length !== arrayOfMaybeDocs.length) {
+                return resolve(
+                  left<CosmosErrors, B>(
+                    toCosmosErrorResponse(
+                      new Error("Some service cannot be decoded correctly")
+                    )
+                  )
+                );
+              } else {
+                rights(arrayOfMaybeDocs).forEach(
+                  (documents: ReadonlyArray<A>) => {
+                    if (documents && documents.length > 0) {
+                      return resolve(
+                        right<CosmosErrors, B>(documents.reduce(f, init))
+                      );
+                    } else {
+                      return resolve(right<CosmosErrors, B>(init));
+                    }
+                  }
+                );
+              }
+            }),
+          reject
+        )
+      )
+  };
+}
+
+async function iteratorToValue<T>(
+  _: IFoldableResultIterator<T>,
+  init: T
+): Promise<Either<CosmosErrors, T>> {
+  async function iterate(a: T): Promise<Either<CosmosErrors, T>> {
+    const errorOrResult = await _.executeNext(a);
+    if (isLeft(errorOrResult)) {
+      return left<CosmosErrors, T>(errorOrResult.value);
+    }
+    const result = errorOrResult.value;
+    return iterate(result);
+  }
+  return iterate(init);
+}
+
```

Comment:
I'm trying to replicate previous logic. Can you check if it's right? @gunzip @cloudify